// @wallet-standard/react-core/lib/cjs/useWallet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWalletProperties = exports.useWallet = exports.WalletContext = void 0;
const react_1 = require("react");
const context_js_1 = require("./context.js");
const DEFAULT_WALLET_PROPERTIES = {
    version: '1.0.0',
    name: context_js_1.EMPTY_STRING,
    icon: `data:image/png;base64,`,
    chains: context_js_1.EMPTY_ARRAY,
    features: context_js_1.EMPTY_OBJECT,
    accounts: context_js_1.EMPTY_ARRAY,
};
const DEFAULT_WALLET_STATE = Object.assign({ wallet: null, 
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    setWallet() { } }, DEFAULT_WALLET_PROPERTIES);
const DEFAULT_WALLET_CONTEXT = (0, context_js_1.createDefaultContext)('Wallet', DEFAULT_WALLET_STATE);
/** TODO: docs */
exports.WalletContext = (0, react_1.createContext)(DEFAULT_WALLET_CONTEXT);
/** TODO: docs */
function useWallet() {
    return (0, react_1.useContext)(exports.WalletContext);
}
exports.useWallet = useWallet;
/** @internal */
function getWalletProperties(wallet) {
    return wallet
        ? {
            version: wallet.version,
            name: wallet.name,
            icon: wallet.icon,
            chains: wallet.chains,
            features: wallet.features,
            accounts: wallet.accounts,
        }
        : DEFAULT_WALLET_PROPERTIES;
}
exports.getWalletProperties = getWalletProperties;
