// @wallet-standard/react-core/lib/esm/useWallet.js
import { createContext, useContext } from 'react';
import { createDefaultContext, EMPTY_ARRAY, EMPTY_OBJECT, EMPTY_STRING } from './context.js';
const DEFAULT_WALLET_PROPERTIES = {
    version: '1.0.0',
    name: EMPTY_STRING,
    icon: `data:image/png;base64,`,
    chains: EMPTY_ARRAY,
    features: EMPTY_OBJECT,
    accounts: EMPTY_ARRAY,
};
const DEFAULT_WALLET_STATE = {
    wallet: null,
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    setWallet() { },
    ...DEFAULT_WALLET_PROPERTIES,
};
const DEFAULT_WALLET_CONTEXT = createDefaultContext('Wallet', DEFAULT_WALLET_STATE);
/** TODO: docs */
export const WalletContext = createContext(DEFAULT_WALLET_CONTEXT);
/** TODO: docs */
export function useWallet() {
    return useContext(WalletContext);
}
/** @internal */
export function getWalletProperties(wallet) {
    return wallet
        ? {
            version: wallet.version,
            name: wallet.name,
            icon: wallet.icon,
            chains: wallet.chains,
            features: wallet.features,
            accounts: wallet.accounts,
        }
        : DEFAULT_WALLET_PROPERTIES;
}
